### PR TITLE
feat(Construct): replace custom resource with bedrock L1 constructs

### LINKED ISSUE #361: (bedrock): adopt official cloudformation resources
### Describe the feature

The Amazon Bedrock team recently released CloudFormation support. https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-bedrock-agent.html

### Use Case

We can remove the bedrock custom resources.

### Proposed Solution

_No response_

### Other Information

_No response_

### Acknowledgements

- [ ] I may be able to implement this feature request
- [X] This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by dineshSajwan**
Fixes #361

**<DO NOT MERGE >**

1. Removed custom resource from bedrock construct.
2. removed custom resource policies from the construct.
3. Upgraded CDK version to 2.141.0
4. Removed knowledge based storage type Redis Enterprise since it is not supported by L1 Constructs
5. Added bedrock L1 construct for datasource , knowledge base, agent and agent alias.
6. Updated test cases for bedrock conctruct
7. Updated agent addAction group and addKnowledgebases method in agent.ts
8. 
---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the project license.

### CODE REVIEW

File: src/cdk-lib/bedrock/agent.ts [hunk 1]
```diff
@@ -587,23 +488,32 @@ export class Agent extends Construct implements cdk.ITaggableV2 {
...
+    };
+
+    this.agent.knowledgeBases= [agentKnowledgeBaseProperty];
+  }
+
+
+  /**
+   * Add action group  to the agent.
+   */
+  public addActionGroups(props: AgentActionGroup) {
```
Comment (krokoko) [inline src/cdk-lib/bedrock/agent.ts:510]: Props should be an array of AgentActionGroup here

File: src/cdk-lib/bedrock/agent.ts [hunk 2]
```diff
@@ -587,23 +488,32 @@ export class Agent extends Construct implements cdk.ITaggableV2 {
...
-    return new AgentActionGroup(this, `AgentActionGroup-${actionGroupName}`, {
-      agent: this,
-      actionGroupName,
-      description: props.description,
-      parentActionGroupSignature: props.parentActionGroupSignature,
-      actionGroupExecutor: props.actionGroupExecutor,
-      actionGroupState: props.actionGroupState,
-      apiSchema: props.apiSchema,
-      shouldPrepareAgent: this.shouldPrepareAgent,
+  public addKnowledgeBase(kb: KnowledgeBase) {
```
Comment (krokoko) [inline src/cdk-lib/bedrock/agent.ts:493]: Props should be an array of kbs and method become ```addKnowledgeBases```

File: src/cdk-lib/bedrock/agent.ts (hunk 1 above)
Comment (dineshSajwan) [reply src/cdk-lib/bedrock/agent.ts]: done - 34db08e7b994c0a4d926f61a9cc9a395a3deba3e

File: src/cdk-lib/bedrock/agent.ts (hunk 2 above)
Comment (dineshSajwan) [reply src/cdk-lib/bedrock/agent.ts]: done - 34db08e7b994c0a4d926f61a9cc9a395a3deba3e

**Comment by krokoko** [comment]
Comments from current testing status:
- tested in TS, deployment of agents with KBs using OSS and Pinecone works as expected
- keep possibility to pass kbs through the construct (call addKBs) 
- Add KBs and add action groups should append to the list, not replace
- Aurora as a vector store fails to deploy:
```
Received response status [FAILED] from custom resource. Message returned: Error: Unable to import module 'custom_resources': No module named 'psycopg2._psycopg
'

Logs: /aws/lambda/***

at invokeUserFunction (/var/task/framework.js:2:6)
at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
at async onEvent (/var/task/framework.js:1:369)
at async Runtime.handler (/var/task/cfn-response.js:1:1676)


Runtime.ImportModuleError: Unable to import module 'custom_resources': No module named 'psycopg2._psycopg'Traceback (most recent call last):
```

File: test/public-registries/PyPi-test.py
Comment (krokoko) [inline test/public-registries/PyPi-test.py:1]: this shouldn't be here

File: test/cdk-lib/bedrock/agent-action-group.test.ts [hunk 3]
```diff
@@ -47,11 +46,12 @@ describe('AgentActionGroup', () => {
       },
     });
 
-    agent = new bedrock.Agent(stack, 'Agent', {
-      foundationModel: bedrock.BedrockFoundationModel.ANTHROPIC_CLAUDE_V2_1,
-      instruction: 'You provide support for developers working with CDK constructs.',
-      idleSessionTTL: cdk.Duration.minutes(30),
-    });
+    // agent = new bedrock.Agent(stack, 'Agent', {
```
Comment (krokoko) [inline test/cdk-lib/bedrock/agent-action-group.test.ts:49]: couple of commented tests in this file

File: test/cdk-lib/bedrock/agent.test.ts [hunk 4]
```diff
@@ -83,35 +86,67 @@ beforeAll(() => {
...
+    description: 'Use these functions to get information about the books in the Project Gutenburg library.',
+    actionGroupState: 'ENABLED',
+    actionGroupExecutor: actionGroupFunction,
+    apiSchema: bedrock.ApiSchema.fromInline('mock schema'),
+  });
+
   agent = new bedrock.Agent(stack, 'Agent', {
     foundationModel: bedrock.BedrockFoundationModel.ANTHROPIC_CLAUDE_V2_1,
     instruction: 'You provide support for developers working with CDK constructs.',
-    knowledgeBases: [kb],
```
Comment (krokoko) [inline test/cdk-lib/bedrock/agent.test.ts:89]: this has been re-added, we can uncomment the line

File: test/cdk-lib/bedrock/agent.test.ts [hunk 5]
```diff
@@ -83,35 +86,67 @@ beforeAll(() => {
...
-  actionGroup = agent.addActionGroup({
-    actionGroupName: 'test-action-group',
-    description: 'Use these functions to get information about the books in the Project Gutenburg library.',
-    actionGroupState: 'ENABLED',
-    apiSchema: bedrock.ApiSchema.fromInline('mock schema'),
-  });
+  agent.addKnowledgeBases([kb]);
+  agent.addActionGroups([actiongroup]);
+
+  // agent.addActionGroup({
```
Comment (krokoko) [inline test/cdk-lib/bedrock/agent.test.ts:129]: commented test

**Review by krokoko — CHANGES_REQUESTED** [review]
- delete the file under 'public-registries'
- couple of commented tests which should be fixed
otherwise should be good !

File: test/public-registries/PyPi-test.py
Comment (dineshSajwan) [reply test/public-registries/PyPi-test.py:1]: removed

File: test/cdk-lib/bedrock/agent-action-group.test.ts (hunk 3 above)
Comment (dineshSajwan) [reply test/cdk-lib/bedrock/agent-action-group.test.ts]: updated - a5b222881d2860c503ed38fcd68ea6619e07542a

File: test/cdk-lib/bedrock/agent.test.ts (hunk 4 above)
Comment (dineshSajwan) [reply test/cdk-lib/bedrock/agent.test.ts]: done - a5b222881d2860c503ed38fcd68ea6619e07542a

File: test/cdk-lib/bedrock/agent.test.ts (hunk 5 above)
Comment (dineshSajwan) [reply test/cdk-lib/bedrock/agent.test.ts]: moved these tests to agent.test.ts - a5b222881d2860c503ed38fcd68ea6619e07542a

**Review by krokoko — APPROVED** [review]

**Review by laithalsaadoon — APPROVED** [review]

**Review by MichaelWalker-git — APPROVED** [review]

**Comment by malikalimoekhamedov** [comment]
> Tested in TS, deployment of agents with KBs using OSS and Pinecone works as expected
> 
> Remaining tasks:
> 
> * [x]  keep possibility to pass kbs through the construct (call addKBs)
> * [x]  Add KBs and add action groups should append to the list, not replace
> * Aurora as a vector store fails to deploy:
> 
> ```
> Received response status [FAILED] from custom resource. Message returned: Error: Unable to import module 'custom_resources': No module named 'psycopg2._psycopg
> '
> 
> Logs: /aws/lambda/***
> 
> at invokeUserFunction (/var/task/framework.js:2:6)
> at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
> at async onEvent (/var/task/framework.js:1:369)
> at async Runtime.handler (/var/task/cfn-response.js:1:1676)
> 
> 
> Runtime.ImportModuleError: Unable to import module 'custom_resources': No module named 'psycopg2._psycopg'�Traceback (most recent call last):
> ```
> 
> This issue is not related to the changes from this branch, and seems to happen on MacOS only. Will open a separate ticket for it
> 
> * [x]  Test and validate Python deployments

@krokoko, what's the separate ticket you've opened? I'm facing this issue and can't deploy my stack with Aurora pgVector.

**Pull request merged**